### ISSUE TITLE
Add -pubout option to genpkey to output the public key

### ISSUE THREAD

**Issue by t8m** [openssl/openssl#22132]
This is needed for algorithms where the public key is not contained within and cannot be generated from the private key.

**Comment by levitte** [comment]
Hmmm...  in other `openssl` commands, `-pubout` is used as a flag to indicate that the file given with `-out` will be populated with the public key (usually as opposed to the private key)...  perhaps another option name should be chosen, for consistency?

**Comment by baentsch** [comment]
> Hmmm... in other `openssl` commands, `-pubout` is used as a flag to indicate that the file given with `-out` will be populated with the public key (usually as opposed to the private key)... perhaps another option name should be chosen, for consistency?

Another option name or another command altogether? Otherwise, this would mean the description of [genpkey](https://www.openssl.org/docs/man3.0/man1/openssl-genpkey.html) needs to change from "generate a private key" to read "generate a public/private key pair", right? Would be in line with what the data structure "(EVP_)PKEY" represents....

**Comment by baentsch** [comment]
From an implementation perspective, it looks rather straightforward: Just add a call to [PEM_write_bio_PUBKEY_ex](https://www.openssl.org/docs/man3.1/man3/PEM_write_bio_PUBKEY_ex.html) -- once we decided on how to specify the filename for the public key, right?

**Comment by levitte** [comment]
> From an implementation perspective, it looks rather straightforward: Just add a call to [PEM_write_bio_PUBKEY_ex](https://www.openssl.org/docs/man3.1/man3/PEM_write_bio_PUBKEY_ex.html) -- once we decided on how to specify the filename for the public key, right?

Yes, that's right, although we might want to consider encoders.

**Comment by levitte** [comment]
> Another option name or another command altogether? Otherwise, this would mean the description of [genpkey](https://www.openssl.org/docs/man3.0/man1/openssl-genpkey.html) needs to change from "generate a private key" to read "generate a public/private key pair", right? Would be in line with what the data structure "(EVP_)PKEY" represents....

I would hate to see a separate command that essentially does the same thing.

You're right about the description, though, it does need to change.  That description emanates from pre-3.0 OpenSSL, where an `EVP_PKEY` was *always* assumed to have the public key (it was generated on the fly by the backend code when necessary, upon retrieval from file).  Had we maintained this requirement with 3.0, all providers would have been required to participate in ensuring that, and this added option would be pretty moot.

**Comment by baentsch** [comment]
> Yes, that's right, although we might want to consider encoders.

What (additional?) consideration would be needed? Do I assume wrong that a call to [PEM_write_bio_PUBKEY_ex](https://www.openssl.org/docs/man3.1/man3/PEM_write_bio_PUBKEY_ex.html) would (automatically) trigger all required encoders?

> I would hate to see a separate command that essentially does the same thing.

Me too. So let's agree to change the description of "genpkey". But the meaning of the "-out" parameter would need to change to read "output the private key" -- maybe only kept as a backwards-compatible option really calling a new "-outprivatekeyfile", made available in addition to a symmetric "-outpublickeyfile" option.

**Comment by t8m** [comment]
I would keep the current -out option as is. And just add another option for the public key output when user asks for that. IMO apart from changing the description of the genpkey command there are no further things to change. If -pubout would be confusing because it is used without an argument in the pkey command, I would call the option -outpubkey. The -outpublickeyfile is IMO too wordy.

**Comment by levitte** [comment]
`-outpubkey` works for me

**Comment by baentsch** [comment]
Can this be closed (as #22173 landed)?

**Comment by levitte** [comment]
Yup
